src/Hamiltonians/HOCartesianEnergyConserved.jl

```diff
@@ -0,0 +1,378 @@
+    aspect::NTuple{D,T}
+    aspect1::NTuple{D,Float64}  # (possibly custom) ratios of trapping frequencies for single-particle energy scale
+    energies::Vector{Float64}   # noninteracting single particle energies
+    vtable::Array{Float64,4}    # interaction coefficients
+    u::Float64
+end
+
+function HOCartesianEnergyConserved(
+        addr::BoseFS; 
+        S = (num_modes(addr),),
+        η = nothing, 
+        g = 1.0,
+        interaction_only = false
+    )
+    # x dimension defines the energy scale, and should be the smallest frequency i.e. largest spatial dimension
+    D = length(S)
+    S_sort = tuple(sort(collect(S); rev = true)...)
+    S == S_sort || @warn("dimensions have been reordered")
+    
+    P = prod(S)
+    P == num_modes(addr) || throw(ArgumentError("number of modes does not match starting address"))
+
+    aspect = box_to_aspect(S)
+    if all(denominator.(aspect) .== 1)
+        aspect = Int.(aspect)
+    end
+
+    if isnothing(η)
+        aspect1 = float.(aspect)
+    elseif length(η) == D
+        aspect1 = S == S_sort ? η : η[sortperm(collect(S); rev = true)]
+    elseif length(η) == 1 && isreal(η) && η ≥ 1
+        aspect1 = ntuple(d -> (d == 1 ? 1.0 : η), D)
+    else
+        throw(ArgumentError("Invalid aspect ratio parameter η."))
+    end
+
+    if interaction_only
+        energies = zeros(P)
+    else
+        states = CartesianIndices(S)    # 1-indexed
+        energies = reshape(map(x -> dot(aspect1, Tuple(x) .- 1/2), states), P)
+    end
+
+    u = sqrt(prod(aspect1)) * g / 2
+
+    max_level = S_sort[1] - 1
+    r = 0:max_level
+    vmat = [four_oscillator_integral_general(i, j, k, l; max_level) for i in r, j in r, k in r, l in r]
+
+    return HOCartesianEnergyConserved{D,typeof(addr),eltype(aspect)}(addr, S_sort, aspect, aspect1, energies, vmat, u)
+end
+
+function Base.show(io::IO, h::HOCartesianEnergyConserved)
+    print(io, "HOCartesianEnergyConserved($(h.addr); S=$(h.S), η=$(h.aspect1), u=$(h.u))")
+end
+
+function starting_address(h::HOCartesianEnergyConserved)
+    return h.addr
+end
+
+LOStructure(::Type{<:HOCartesianEnergyConserved}) = IsHermitian()
+
+
+### DIAGONAL ELEMENTS ###
+function energy_transfer_diagonal(h::HOCartesianEnergyConserved{D}, omm::BoseOccupiedModeMap) where {D}
+    result = 0.0
+    states = CartesianIndices(h.S)    # 1-indexed
+    
+    for i in eachindex(omm)
+        mode_i, occ_i = omm[i].mode, omm[i].occnum
+        idx_i = Tuple(states[mode_i])
+        if occ_i > 1
+            # use i in 1:M indexing for accessing table
+            val = occ_i * (occ_i - 1)
+            result += prod(h.vtable[idx_i[d],idx_i[d],idx_i[d],idx_i[d]] for d in 1:D) * val
+        end
+        for j in 1:i-1
+            mode_j, occ_j = omm[j].mode, omm[j].occnum
+            idx_j = Tuple(states[mode_j])
+            val = 4 * occ_i * occ_j
+            result += prod(h.vtable[idx_i[d],idx_j[d],idx_i[d],idx_j[d]] for d in 1:D) * val
+        end        
```

Comment:
Tests don't enter this for-loop. Should we worry?
https://coveralls.io/builds/59714810/source?filename=src%2FHamiltonians%2FHOCartesianEnergyConserved.jl#L236